ftw/monitor/perf_metrics.py

```diff
@@ -0,0 +1,105 @@
+from ftw.monitor.server import get_uptime
+from time import time
+from ZODB.interfaces import IDatabase
+from zope.component import getUtility
+from Zope2 import app as App  # noqa
+import json
+import psutil
+
+
+def perf_metrics(connection, database='-', sampling_interval=300):
+    r"""Get performance related metrics.
+
+    Usage: Send the message 'perf_metrics\r\n' to the zc.monitor TCP port.
+
+    The plugin will then respond with a JSON encoded dictionary containing
+    several performance related metrics.
+
+    You can pass a database name, where "-" is an alias for the main database.
+
+    By default, the statistics are for a sampling interval of 5
+    minutes.  You can request another sampling interval, up to an
+    hour, by passing a sampling interval in seconds after the database name.
+    The sampling interval applies to statistics tracked by the ZODB
+    ActivityMonitor.
+    """
+    app = App()
+
+    # Make sure we have a connected storage available
+    try:
+        dbchooser = app.Control_Panel.Database
+        for dbname in dbchooser.getDatabaseNames():
+            storage = dbchooser[dbname]._getDB()._storage
+            is_connected = getattr(storage, 'is_connected', None)
+            if is_connected is not None and not is_connected():
+                msg = {'error': 'Database %r disconnected.' % dbname}
+                connection.write('%s\n' % json.dumps(msg))
+                return
+
+        # Get DB. These will be registered during startup in f.m.server
+        if database == '-':
+            database = 'main'
+        db = getUtility(IDatabase, database)
+
+        metrics = collect_performance_metrics(db, sampling_interval)
+        connection.write('%s\n' % json.dumps(metrics))
+
+    finally:
+        app._p_jar.close()
+
+
+def collect_performance_metrics(db, sampling_interval):
+    instance_stats = get_instance_stats()
+    cache_stats = get_cache_stats(db)
+    db_stats = get_db_stats(db, sampling_interval)
+    memory_stats = get_memory_stats()
+    metrics = {
+        'instance': instance_stats,
+        'cache': cache_stats,
+        'db': db_stats,
+        'memory': memory_stats,
+    }
+    return metrics
+
+
+def get_instance_stats():
+    return {'uptime': int(get_uptime())}
```

Comment:
BTW, I've seen you're storing the start time yourself. It's also available from `app.Control_Panel.process_start`.

Reply:
Yes. I originally needed to track start time myself because `process_start` wasn't available early enough in the application lifecycle (it's needed just before warmup, to avoid flickering of the healthy state).

I've briefly looked into using [`ApplicationManager`](https://github.com/zopefoundation/Zope/blob/12a4c8826448d7cde245de205c9514128ae86a9c/src/App/ApplicationManager.py#L135)`.process_time`, but wasn't quite comfortable using a persistent object across different DB connections, so I used the uptime that's already tracked in `ftw.monitor`.